xmtp/src/client.rs

```diff
@@ -34,71 +42,110 @@ where
     pub persistence: NamespacedPersistence<P>,
     pub(crate) account: Account,
     pub(super) _store: S,
-}
 
-impl<A, P, S> Client<A, P, S>
-where
-    A: XmtpApiClient,
-    P: Persistence,
-{
+    is_initialized: bool,
 }
 
 impl<A, P, S> Client<A, P, S>
 where
     A: XmtpApiClient,
     P: Persistence,
 {
-    pub fn write_to_persistence(&mut self, s: &str, b: &[u8]) -> Result<(), P::Error> {
-        self.persistence.write(s, b)
-    }
-
-    pub fn read_from_persistence(&self, s: &str) -> Result<Option<Vec<u8>>, P::Error> {
-        self.persistence.read(s)
+    pub fn new(
+        api_client: A,
+        network: Network,
+        persistence: NamespacedPersistence<P>,
+        account: Account,
+        store: S,
+    ) -> Self {
+        Self {
+            api_client,
+            network,
+            persistence,
+            account,
+            _store: store,
+            is_initialized: false,
+        }
     }
 
     pub fn wallet_address(&self) -> Address {
         self.account.addr()
     }
 
-    pub async fn get_contacts(
+    pub async fn init(&mut self) -> Result<(), ClientError> {
+        // Register Contact Bundles
+        let registered_bundles = self.get_contacts(&self.wallet_address()).await?;
+
+        if registered_bundles.is_empty() {
+            self.publish_user_contact().await?;
+        }
+
+        self.is_initialized = true;
+        Ok(())
+    }
+
+    async fn get_contacts(
         &self,
         wallet_address: &str,
-    ) -> Result<Vec<VmacContactBundle>, String> {
+    ) -> Result<Vec<VmacContactBundle>, ClientError> {
         let topic = build_user_contact_topic(wallet_address.to_string());
-        let response = self.api_client.query(topic, None, None, None).await?;
+        let response = self
+            .api_client
+            .query(topic, None, None, None)
+            .await
+            .map_err(ClientError::QueryError)?;
 
         let mut contacts = vec![];
         for envelope in response.envelopes {
             // TODO: Wrap the proto in some special struct
             // TODO: Handle errors better
-            let contact_bundle = VmacContactBundle::decode(envelope.message.as_slice())
-                .map_err(|e| format!("{}", e))?;
+            let contact_bundle =
```

Comment:
I foresee some merge conflicts with what I'm working on. I have a new `Contact` struct in my branch. I'll work out the conflicts on my end, but I'll probably steal some of your error handling code to reduce the burden.

Reply:
Sorry, beat you to the punch merging in the `Contact` struct, which'll mean you'll have to deal with a few merge conflicts.